### ISSUE TITLE
[Regression][RedHat69] System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandler_SslProtocols_Test / GetAsync_AllowedSSLVersion_Succeeds

### ISSUE THREAD

**Issue by karelz** [dotnet/corefx#28790]
## Type of failures

Test case: `acceptedProtocol: Ssl3, requestOnlyThisProtocol: True`

```
System.Net.Http.HttpRequestException: An error occurred while sending the request.
----   System.Net.Http.CurlException : SSL connect error
at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask,   HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)   in /root/corefx-1549868/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line   479
at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<>c__DisplayClass4_0.<<GetAsync_AllowedSSLVersion_Succeeds>b__0>d.MoveNext()   in /root/corefx-1549868/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.SslProtocols.cs:line   126
--- End of stack trace from previous location where exception was thrown   ---
at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2   funcAsync, Options options) in   /root/corefx-1549868/src/Common/tests/System/Net/Http/LoopbackServer.cs:line   67
at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds(SslProtocols   acceptedProtocol, Boolean requestOnlyThisProtocol) in   /root/corefx-1549868/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.SslProtocols.cs:line   124
--- End of stack trace from previous location where exception was thrown   ---
----- Inner Stack Trace -----
at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error) in   /root/corefx-1549868/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.cs:line   660
at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1   easyWrapper, CURLcode messageResult) in /root/corefx-1549868/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs:line   883
```

## History of failures

Day | Build | Details
-- | -- | --
3/31 | 20180331.05 | 
4/1 | 20180401.01 | 
4/1 | 20180401.02 | 
4/2 | 20180402.01 | 
4/2 | 20180402.02 | 
4/2 | 20180402.05 | 
4/3 | 20180403.01 | 
4/3 | 20180403.04 | 

All test cases are `acceptedProtocol: Ssl3, requestOnlyThisProtocol: True`

**Comment by stephentoub** [comment]
@karelz, with what inputs?  GetAsync_AllowedSSLVersion_Succeeds is a theory.

**Comment by karelz** [comment]
All test cases are `acceptedProtocol: Ssl3, requestOnlyThisProtocol: True` - I assume it is fallout from Ssl2/Ssl3 being now enabled.
I assume we should just disable the test case on RedHat69 as it is not supported there at all ...

**Comment by stephentoub** [comment]
@tmds, would you expect SSL3 to work on RedHat69, or was support for it even then removed?

**Comment by karelz** [comment]
@rmkerr can you please take look?

**Comment by rmkerr** [comment]
Taking a look at this now.

**Comment by tmds** [comment]
> @tmds, would you expect SSL3 to work on RedHat69, or was support for it even then removed?

@stephentoub @rmkerr I have no idea.
The test is checking this condition:
https://github.com/dotnet/corefx/blob/3cc6fda0c6abe76eea222c165ebdf816e83df564/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.SslProtocols.Unix.cs#L19-L21
On RH7 the backend is NSS, so this test will be skipped. Is RH6 using libcurl with openssl backend by default? Or is there a custom libcurl package on the CI machines?

**Comment by rmkerr** [comment]
On one of the RH6 CI machines `curl -V` gives me the following output:
```
curl 7.45.0 (x86_64-pc-linux-gnu) libcurl/7.45.0 OpenSSL/1.0.1e zlib/1.2.3 libidn/1.18 libssh2/1.4.2 nghttp2/1.6.0
Protocols: http https scp sftp
Features: IDN IPv6 Largefile NTLM NTLM_WB SSL libz HTTP2 UnixSockets
```

But in the default RH6 install I can see that an earlier version is installed:
```
curl 7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.27.1 zlib/1.2.3 libidn/1.18 libssh2/1.4.2
Protocols: tftp ftp telnet dict ldap ldaps http file https ftps scp sftp
Features: GSS-Negotiate IDN IPv6 Largefile NTLM SSL libz
```
I'm still learning about how we use libcurl and openssl, so I don't know what kind of consequences this would have. Any thoughts @tmds?

**Comment by karelz** [comment]
Remaining investigation is not critical for 2.1, moving to Future.

**Comment by tmds** [comment]
@karelz it would be nice to know the root cause, but I don't think it will buy us something.

**Comment by wfurt** [comment]
This is failing on Windows.10.Amd64.Core now. 24 failures today. I think we need better detection if SSL3 should work or not. 

```
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in F:\vsagent\1\s\src\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs:line 105
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in F:\vsagent\1\s\src\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs:line 65
   at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.<>c__DisplayClass4_0.<<GetAsync_AllowedSSLVersion_Succeeds>b__0>d.MoveNext() in F:\vsagent\1\s\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs:line 144
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in F:\vsagent\1\s\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 66
   at System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.GetAsync_AllowedSSLVersion_Succeeds(SslProtocols acceptedProtocol, Boolean requestOnlyThisProtocol) in F:\vsagent\1\s\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.SslProtocols.cs:line 142
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace #1 (System.AggregateException) -----

----- Inner Stack Trace -----
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 879
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 868
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 661
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 580
   at System.Net.Security.SslStream.BeginAuthenticateAsServer(SslServerAuthenticationOptions sslServerAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 266
   at System.Net.Security.SslStream.BeginAuthenticateAsServer(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 255
   at System.Net.Security.SslStream.<>c.<AuthenticateAsServerAsync>b__50_1(X509Certificate arg1, Boolean arg2, SslProtocols arg3, AsyncCallback callback, Object state) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 418
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions) in F:\vsagent\15\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\FutureFactory.cs:line 1158
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions) in F:\vsagent\15\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\TaskFactory.cs:line 1135
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state) in F:\vsagent\15\s\src\System.Private.CoreLib\shared\System\Threading\Tasks\TaskFactory.cs:line 1092
   at System.Net.Security.SslStream.AuthenticateAsServerAsync(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation) in F:\vsagent\1\s\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 419
   at System.Net.Test.Common.LoopbackServer.AcceptConnectionAsync(Func`2 funcAsync) in F:\vsagent\1\s\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 100
   at System.Net.Test.Common.LoopbackServer.AcceptConnectionSendResponseAndCloseAsync(HttpStatusCode statusCode, String additionalHeaders, String content) in F:\vsagent\1\s\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 146
----- Inner Stack Trace #2 (System.AggregateException) -----

----- Inner Stack Trace -----
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in F:\vsagent\1\s\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 475
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in F:\vsagent\1\s\src\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs:line 77
----- Inner Stack Trace -----
   at System.Threading.Tasks.RendezvousAwaitable`1.GetResult() in F:\vsagent\1\s\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs:line 66
   at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state) in F:\vsagent\1\s\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 880
```

**Comment by danmosemsft** [comment]
This is basically failing all the time now. @wfurt any chance you could either disable or take a shot at fixing it? 🐱 

https://mc.dot.net/#/product/netcore/30/source/official~2Fdotnet~2Fcorefx~2Frefs~2Fheads~2Fmaster/type/test~2Ffunctional~2Fcli~2F/build/20190206.6/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandler_SslProtocols_Test~2FGetAsync_AllowedSSLVersion_Succeeds(acceptedProtocol:%20Ssl3,%20requestOnlyThisProtocol:%20True)

**Comment by danmosemsft** [comment]
here's the message too
```
System.AggregateException : One or more errors occurred. (One or more errors occurred. (Authentication failed because the remote party has closed the transport stream.)) (One or more errors occurred. (An error occurred while sending the request.))
---- System.AggregateException : One or more errors occurred. (Authentication failed because the remote party has closed the transport stream.)
-------- System.IO.IOException : Authentication failed because the remote party has closed the transport stream.
---- System.AggregateException : One or more errors occurred. (An error occurred while sending the request.)
-------- System.Net.Http.HttpRequestException : An error occurred while sending the request.
------------ System.Net.Http.WinHttpException : Error 12175 calling WINHTTP_CALLBACK_STATUS_REQUEST_ERROR, 'A security error occurred'.
```

**Comment by stephentoub** [comment]
When did it start failing?

**Comment by stephentoub** [comment]
Oops, nevermind.

**Comment by danmosemsft** [comment]
Oh I see @baulig has https://github.com/dotnet/corefx/pull/35250 ... I guess that will fix?

**Comment by wfurt** [comment]
We disabled test on Redhat6 and now t is failing cross the board as distros (and Windows) are dropping SSLv3 support. And yes, #35250 should fix it. 
Alternatively since Ssl3 is practically dead, we may retire it from test set.

**Comment by wfurt** [comment]
closed by #35574
